### PR TITLE
ICM fixes (5/n)

### PR DESCRIPTION
**Pull request by hariharans29**
### Description
Fixes 3 ICM fixes:

https://portal.microsofticm.com/imp/v5/incidents/details/31000000572208/summary
https://portal.microsofticm.com/imp/v5/incidents/details/31000000573313/summary
https://portal.microsofticm.com/imp/v5/incidents/details/31000000575583/summary

### Motivation and Context
Fix ICM issues

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h [hunk 1]
```diff
@@ -73,20 +80,45 @@ Status CheckInputs(const T* input,
...
-                                            : static_cast<int>(hidden_size / num_heads);
+  int max_sequence_length = 0;
+  int head_size = 0;
+  ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(cos_cache_dims[0], "max_sequence_length", max_sequence_length));
+  if (rotary_embedding_dim == 0) {
+    int cache_width = 0;
+    ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(cos_cache_dims[1], "cache_width", cache_width));
+    ORT_RETURN_IF_ERROR(detail::CheckedMulToInt32(cache_width, 2, "head_size", head_size));
+  } else {
+    if (!transposed && hidden_size % num_heads != 0) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h]: ⚠️ **Division-by-zero when `num_heads == 0`**: This `hidden_size % num_heads` can be undefined behavior if `num_heads` is 0 (its default attribute value). The equivalent code in the llm variant (`core/providers/cpu/llm/rotary_embedding_helper.h`) correctly adds a `num_heads <= 0` guard before any division:

```cpp
if (num_heads <= 0) {
  return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                         "RotaryEmbedding: num_heads must be greater than 0 for rank-3 input");
}
```

The same guard should be added here before the modulo check, for rank-3 input when `rotary_embedding_dim > 0`.

**Review by tianleiwu — COMMENTED** [review]
## Review Summary

Solid hardening PR — the checked arithmetic approach (`NarrowNonNegativeToInt32`, `CheckedMulToInt32`, `CheckedMulToPtrdiff`) is thorough and well-structured. CumSum empty-axis fix and LinearRegressor validations are clean. Test coverage is good across all three fix areas.

One high-priority concern: the contrib RotaryEmbedding helper has a potential division-by-zero that the llm variant correctly guards against (see inline comment).

File: onnxruntime/contrib_ops/cpu/bert/rotary_embedding.cc [hunk 2]
```diff
@@ -59,6 +112,9 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
   const int rotary_emb_dim = parameters.rotary_embedding_dim;
   const int half_rotary_emb_dim = rotary_emb_dim / 2;
 
+  std::ptrdiff_t position_count = 0;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/rotary_embedding.cc:115]: How about using SafeInt directly
```
std::ptrdiff_t position_count = SafeInt<std::ptrdiff_t>(batch_size) * sequence_length;
```

SafeInt will call ORT_THROW if there is overflow.

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc [hunk 3]
```diff
@@ -49,9 +103,11 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
   const int rotary_emb_dim = parameters.rotary_embedding_dim;
   const int half_rotary_emb_dim = rotary_emb_dim / 2;
 
+  std::ptrdiff_t position_count = SafeInt<std::ptrdiff_t>(batch_size) * sequence_length;
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/llm/rotary_embedding.cc]: `SafeInt` overflows call `ORT_THROW`, and `ORT_THROW` aborts in `ORT_NO_EXCEPTIONS` builds. That means malformed shapes can still terminate the process here before the later checked `total_elements` path gets a chance to return `INVALID_ARGUMENT`. Since `position_count` is only needed for `position_ids_format == 1`, could we keep this on the `Status` path instead, e.g. compute it lazily with `CheckedMulToPtrdiff` inside that branch?

File: onnxruntime/contrib_ops/cpu/bert/rotary_embedding.cc [hunk 4]
```diff
@@ -59,6 +113,8 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
   const int rotary_emb_dim = parameters.rotary_embedding_dim;
   const int half_rotary_emb_dim = rotary_emb_dim / 2;
 
+  std::ptrdiff_t position_count = SafeInt<std::ptrdiff_t>(batch_size) * sequence_length;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/rotary_embedding.cc]: Same issue as the ONNX RotaryEmbedding kernel: `SafeInt` overflow routes through `ORT_THROW`, which aborts in `ORT_NO_EXCEPTIONS` builds. That reintroduces a hard-failure path for malformed shapes in a codepath this PR is otherwise converting to clean validation errors. Could we switch this back to a `Status`-returning checked multiply and only evaluate it when `position_ids_format == 1`?

**Review by tianleiwu — CHANGES_REQUESTED** [review]
The hardening work here is solid overall, especially the checked narrowing in RotaryEmbedding and the new negative tests across RotaryEmbedding, CumSum, and LinearRegressor. I found one remaining issue in the rotary overflow path: the latest `position_count` change now goes through `SafeInt`, which can still terminate `ORT_NO_EXCEPTIONS` builds instead of returning a normal `INVALID_ARGUMENT` status for malformed shapes.

File: onnxruntime/contrib_ops/cpu/attnlstm/uni_dir_attn_lstm.cc [hunk 5]
```diff
@@ -216,7 +216,7 @@ void UniDirectionalAttnLstm<T>::LoadPeepholeWeights(const gsl::span<const T>& pe
 template <typename T>
 void UniDirectionalAttnLstm<T>::LoadBias(const gsl::span<const T>& WbRb_values) {
   const size_t hidden_size = CheckedToSizeT(hidden_size_);
-  const size_t Wb_to_Rb_offset = CheckedMulToSizeT(hidden_size_, 4);
+  const size_t Wb_to_Rb_offset = CheckedMulToSizeT(hidden_size, size_t{4});
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cpu/attnlstm/uni_dir_attn_lstm.cc:219]: https://github.com/microsoft/onnxruntime/pull/27906#issuecomment-4282649255

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc [hunk 6]
```diff
@@ -62,7 +118,23 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
...
+  std::ptrdiff_t max_cache_offset = 0;
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(batch_size - 1, 0), batch_stride, "max_batch_offset", max_batch_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(sequence_length - 1, 0), seq_stride, "max_seq_offset", max_seq_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(n_heads - 1, 0), head_stride, "max_head_offset", max_head_offset));
+  ORT_RETURN_IF_ERROR(CheckedAddToPtrdiff(max_batch_offset, max_seq_offset, "max_block_offset", max_block_offset));
+  ORT_RETURN_IF_ERROR(CheckedAddToPtrdiff(max_block_offset, max_head_offset, "max_block_offset", max_block_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(max_sequence_length - 1, 0), half_rotary_emb_dim,
+                                          "max_cache_offset", max_cache_offset));
+  static_cast<void>(max_cache_offset);
+
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/llm/rotary_embedding.cc:138]: _(Moved to inline comment)_

File: onnxruntime/test/providers/cpu/ml/linearregressor_test.cc [hunk 7]
```diff
@@ -98,5 +100,36 @@ INSTANTIATE_TEST_SUITE_P(
...
+  OpTester test("LinearRegressor", 1, onnxruntime::kMLDomain);
+
+  test.AddAttribute("targets", std::numeric_limits<int64_t>::max());
+  test.AddAttribute("coefficients", std::vector<float>{1.f});
+  test.AddAttribute("intercepts", std::vector<float>{0.f});
+
+  test.AddInput<float>("X", {1, 1}, {1.f});
+  test.AddOutput<float>("Y", {1, 1}, {0.f});
+
+  test.Run(OpTester::ExpectResult::kExpectFailure, "targets must be in range [1,");
```
Comment (tianleiwu) [inline onnxruntime/test/providers/cpu/ml/linearregressor_test.cc:127]: _(Moved to inline comment)_

**Review by tianleiwu — DISMISSED** [review]
**Overall**: Thorough, systematic hardening of integer arithmetic across four operators (RotaryEmbedding, CumSum, LinearRegressor, AttnLSTM). All `static_cast<int>` narrowing replaced with checked helpers, dimension products validated before pointer arithmetic, loop indices widened to `ptrdiff_t`. Comprehensive regression tests cover each fix. Previous automated-review concerns (30 threads) have all been addressed in this head.

Remaining items are minor quality-of-life improvements, none blocking.

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc [hunk 8]
```diff
@@ -62,7 +118,23 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
...
+  std::ptrdiff_t max_block_offset = 0;
+  std::ptrdiff_t max_cache_offset = 0;
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(batch_size - 1, 0), batch_stride, "max_batch_offset", max_batch_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(sequence_length - 1, 0), seq_stride, "max_seq_offset", max_seq_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(n_heads - 1, 0), head_stride, "max_head_offset", max_head_offset));
+  ORT_RETURN_IF_ERROR(CheckedAddToPtrdiff(max_batch_offset, max_seq_offset, "max_block_offset", max_block_offset));
+  ORT_RETURN_IF_ERROR(CheckedAddToPtrdiff(max_block_offset, max_head_offset, "max_block_offset", max_block_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(max_sequence_length - 1, 0), half_rotary_emb_dim,
+                                          "max_cache_offset", max_cache_offset));
+  static_cast<void>(max_cache_offset);
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/llm/rotary_embedding.cc:137]: Suggestion: `max_cache_offset` is computed but only used as a validation side-effect, immediately suppressed with `static_cast<void>`. This pattern is correct (validates the offset fits `ptrdiff_t`) but is unusual in this codebase. A brief comment like `// validate max cache offset fits in ptrdiff_t` would clarify intent and prevent accidental removal.

Also: this validates `(max_sequence_length - 1) * half_rotary_emb_dim`, which transitively protects the `cache_offset = position_id * half_rotary_emb_dim` multiplication inside the loop body (since position_ids are range-checked against `max_sequence_length`). That transitive relationship is non-obvious — a linking comment in the loop body would help.

File: onnxruntime/test/providers/cpu/ml/linearregressor_test.cc [hunk 9]
```diff
@@ -98,5 +100,36 @@ INSTANTIATE_TEST_SUITE_P(
...
+
+  test.AddInput<float>("X", {1, 2}, {1.f, 2.f});
+  test.AddOutput<float>("Y", {1, 1}, {0.f});
+
+  test.Run(OpTester::ExpectResult::kExpectFailure, "LinearRegressor: coefficients length (1) must equal targets (1) * features (2)");
+}
+
+TEST(MLOpTest, LinearRegressorOversizedTargets) {
+  if (std::numeric_limits<int64_t>::max() <= std::numeric_limits<std::ptrdiff_t>::max()) {
+    GTEST_SKIP() << "No int64_t value exceeds ptrdiff_t on this platform.";
```
Comment (tianleiwu) [inline onnxruntime/test/providers/cpu/ml/linearregressor_test.cc]: Nitpick: `LinearRegressorOversizedTargets` gates on `std::numeric_limits<int64_t>::max() > std::numeric_limits<std::ptrdiff_t>::max()`, which is false on all standard 64-bit platforms (both are the same type). This means the test effectively never runs on Linux/macOS/Windows x64 builds. The constructor `ORT_ENFORCE` is still reachable via other paths, but this specific regression test is dead code on common platforms.

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc [hunk 10]
```diff
@@ -62,7 +118,23 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
...
+
   // The cost is calculated as:
   //   - head_size * sizeof(T) for reading input
   //   - head_size * sizeof(T) for writing output
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/llm/rotary_embedding.cc:141]: _(Moved to inline comment)_

File: onnxruntime/test/providers/cpu/ml/linearregressor_test.cc [hunk 11]
```diff
@@ -98,5 +100,36 @@ INSTANTIATE_TEST_SUITE_P(
...
+  }
+
+  OpTester test("LinearRegressor", 1, onnxruntime::kMLDomain);
+
```
Comment (tianleiwu) [inline onnxruntime/test/providers/cpu/ml/linearregressor_test.cc:119]: _(Moved to inline comment)_

**Review by tianleiwu — COMMENTED** [review]
_(Duplicate comments — see inline comments from the other review thread)_

**Review by tianleiwu — COMMENTED** [review]
## Review Summary

Well-executed security hardening. The checked-arithmetic approach (`NarrowNonNegativeToInt32`, `CheckedMulToInt32`, `CheckedMulToPtrdiff`) is thorough, and the pre-loop maximum-offset validation pattern in `RunRotaryEmbedding` is elegant — it proves all inner-loop pointer arithmetic is safe without per-iteration checks. Test coverage is solid with platform-aware overflow tests. Two inline suggestions below, neither blocking.

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc [hunk 12]
```diff
@@ -12,6 +15,48 @@ using namespace onnxruntime::rotary_embedding_helper;
 
 namespace onnxruntime {
 
+namespace {
+
+inline Status CheckedMulToPtrdiff(int lhs, int rhs, const char* name, std::ptrdiff_t& output) {
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/llm/rotary_embedding.cc]: Suggestion: `CheckedMulToPtrdiff` (2-arg and 3-arg overloads) and `CheckedAddToPtrdiff` are identical between this file and `contrib_ops/cpu/bert/rotary_embedding.cc`. Since `rotary_embedding_int32_utils.h` was already created to share the int32 helpers, consider extracting these ptrdiff helpers into a shared header as well (e.g., extend `rotary_embedding_int32_utils.h` or create `rotary_embedding_ptrdiff_utils.h`). If a bug is found in one copy, the other might not get updated.

File: onnxruntime/core/providers/cpu/ml/linearregressor.cc [hunk 13]
```diff
@@ -94,9 +104,9 @@ Status LinearRegressor::Compute(OpKernelContext* ctx) const {
                            "num_targets (", num_targets_, ") * num_features (", num_features,
                            ") overflows size_t");
   }
-  ORT_RETURN_IF_NOT(coefficients_.size() >= expected_coefficients_size,
-                    "coefficients size (", coefficients_.size(), ") is less than num_targets (", num_targets_,
-                    ") * num_features (", num_features, ")");
+  ORT_RETURN_IF_NOT(coefficients_.size() == expected_coefficients_size,
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Suggestion: This tightens the check from `>= expected_coefficients_size` to `== expected_coefficients_size`. The old behavior silently accepted models with excess coefficients; the new strict equality will reject them at runtime. While arguably more correct (excess coefficients likely indicate a model authoring error), this is a behavior change that could break existing deployed models that happen to have extra coefficients. Consider whether this strictness change is intentional for the ICM fix, or whether keeping `>=` with an added warning log would be safer.

**Review by tianleiwu — COMMENTED** [review]
## Review Summary

Well-executed security hardening. The checked-arithmetic approach (`NarrowNonNegativeToInt32`, `CheckedMulToInt32`, `CheckedMulToPtrdiff`) is thorough, and the pre-loop maximum-offset validation pattern in `RunRotaryEmbedding` is elegant — it proves all inner-loop pointer arithmetic is safe without per-iteration checks. Test coverage is solid with platform-aware overflow tests. Two inline suggestions below, neither blocking.

File: onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h [hunk 14]
```diff
@@ -73,20 +80,49 @@ Status CheckInputs(const T* input,
...
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "RotaryEmbedding: num_heads must be greater than 0 for rank-3 input");
+    }
+    if (!transposed && hidden_size % num_heads != 0) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "RotaryEmbedding: hidden_size=", hidden_size,
+                             " must be divisible by num_heads=", num_heads,
+                             " for rank-3 input");
+    }
+    head_size = static_cast<int>(hidden_size / num_heads);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h:107]: **Suggestion: Guard against `num_heads == 0` for rank-4 transposed input**

The `!transposed` guards on lines 110 and 114 protect rank-3 inputs from dividing by zero, but for rank-4 transposed input with `rotary_embedding_dim != 0`, this line still executes `hidden_size / num_heads` where `num_heads` is the attribute value (could be 0). This is a pre-existing issue, but since this PR is hardening this exact code path, it would be natural to fix it:

```cpp
if (num_heads == 0) {
  head_size = 0;  // empty or degenerate tensor
} else {
  if (!transposed && hidden_size % num_heads != 0) {
    return ORT_MAKE_STATUS(...);
  }
  head_size = static_cast<int>(hidden_size / num_heads);
}
```

Alternatively, for rank-4 transposed input, `head_size` is already known from `input_dims[3]` — it could be preserved from the rank-4 branch instead of recomputing via division.

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc [hunk 15]
```diff
@@ -62,7 +118,24 @@ Status RunRotaryEmbedding(concurrency::ThreadPool* tp, RotaryParameters paramete
...
+  std::ptrdiff_t max_cache_offset = 0;
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(batch_size - 1, 0), batch_stride, "max_batch_offset", max_batch_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(sequence_length - 1, 0), seq_stride, "max_seq_offset", max_seq_offset));
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(n_heads - 1, 0), head_stride, "max_head_offset", max_head_offset));
+  ORT_RETURN_IF_ERROR(CheckedAddToPtrdiff(max_batch_offset, max_seq_offset, "max_block_offset", max_block_offset));
+  ORT_RETURN_IF_ERROR(CheckedAddToPtrdiff(max_block_offset, max_head_offset, "max_block_offset", max_block_offset));
+  // Validate that the largest cache lookup offset fits in ptrdiff_t.
+  ORT_RETURN_IF_ERROR(CheckedMulToPtrdiff(std::max(max_sequence_length - 1, 0), half_rotary_emb_dim,
+                                          "max_cache_offset", max_cache_offset));
+  static_cast<void>(max_cache_offset);
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/llm/rotary_embedding.cc]: Nitpick: `[[maybe_unused]]` on the variable declaration would be more idiomatic than `static_cast<void>`:
```cpp
[[maybe_unused]] std::ptrdiff_t max_cache_offset = 0;
```
Same applies to the identical pattern in `contrib_ops/cpu/bert/rotary_embedding.cc:159`.

**Review by tianleiwu — COMMENTED** [review]
## Review Summary

Solid hardening PR. The systematic replacement of unsafe `static_cast<int>()` narrowing with overflow-checked helpers (`NarrowNonNegativeToInt32`, `CheckedMulToInt32`, `CheckedMulToPtrdiff`) is thorough and well-tested. Widening loop indices/offsets to `ptrdiff_t` and pre-validating max offsets before pointer arithmetic is good defense-in-depth.

**Highlights:**
- Clean shared utility in `rotary_embedding_int32_utils.h` with specific error messages
- Consistent application across both ONNX and contrib RotaryEmbedding variants
- Good test coverage: overflow, divisibility, empty axis, invalid targets
- AttnLSTM type fix (`hidden_size` → `size_t{4}`) is a clean correctness improvement

**Two suggestions below** — neither is blocking.

File: onnxruntime/core/providers/cpu/llm/rotary_embedding.cc (hunk 12 above)
Comment (tianleiwu) [reply onnxruntime/core/providers/cpu/llm/rotary_embedding.cc]: +1 on this. The ptrdiff_t helpers (`CheckedMulToPtrdiff` 2-arg/3-arg, `CheckedAddToPtrdiff`) are now defined identically in three places: `rotary_embedding_int32_utils.h`, the anonymous namespace here, and the anonymous namespace in `contrib_ops/cpu/bert/rotary_embedding.cc`. These .cc files could simply import from the shared header, matching the pattern already used for `NarrowNonNegativeToInt32` and `CheckedMulToInt32`.

File: onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h [hunk 16]
```diff
@@ -73,20 +80,50 @@ Status CheckInputs(const T* input,
...
+  ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(input_dims[2], "hidden_size", hidden_size));
 
   bool transposed = false;
   if (input_dims.size() == 4) {
     // input is [batch, num_heads, seq, head_size]
-    sequence_length = static_cast<int>(input_dims[2]);
-    hidden_size = static_cast<int>(input_dims[1]) * static_cast<int>(input_dims[3]);
+    int input_num_heads = 0;
+    ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(input_dims[2], "sequence_length", sequence_length));
+    ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(input_dims[1], "num_heads", input_num_heads));
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h]: Suggestion (non-blocking): For rank-4 input, `input_dims[1]` is stored in a local `input_num_heads` rather than directly updating the function parameter `num_heads`. This works correctly — `hidden_size = input_num_heads * head_size` is computed, and `num_heads` is later inferred as `hidden_size / head_size`. But the ONNX version (`core/providers/cpu/llm/rotary_embedding_helper.h`) directly sets `num_heads` from `input_dims[1]` for rank-4 input. Aligning the two would make the contrib code easier to reason about.

**Review by tianleiwu — DISMISSED** [review]
## Review Summary

Well-executed security hardening. The systematic replacement of unchecked `static_cast<int>()` narrowing with `NarrowNonNegativeToInt32` and `CheckedMulToInt32` helpers is thorough and addresses real integer overflow vulnerabilities in RotaryEmbedding, CumSum, and LinearRegressor.

**Highlights:**
- `rotary_embedding_int32_utils.h` cleanly centralizes checked arithmetic. All helpers return `INVALID_ARGUMENT` status, properly classifying errors as input validation failures.
- The overflow checks (`lhs != 0 && rhs > max / lhs`) are correct for non-negative inputs.
- Position ID bounds checking (both format 0 and format 1) is now done before the parallel loop.
- CumSum: `axis_tensor->Shape().Size() != 1` closes a real gap where an empty 1-D axis tensor would read OOB memory.
- LinearRegressor: `num_targets_ > 0` and 0-dimensional input checks catch invalid models early.
- The `uni_dir_attn_lstm.cc` fix correctly uses the already-narrowed `hidden_size` (type `size_t`) instead of the raw `hidden_size_` (`int`) in `CheckedMulToSizeT`.
- Tests effectively cover overflow, divisibility, and empty-tensor edge cases.

**One minor suggestion (not blocking):** In `contrib_ops/cpu/bert/rotary_embedding_helper.h`, for rank-4 input, `input_dims[1]` is stored in a local `input_num_heads` rather than directly updating the `num_heads` parameter. This works correctly (the value is later inferred from `hidden_size / head_size`), but the ONNX helper directly sets `num_heads` from `input_dims[1]`. The inconsistency makes the contrib code slightly harder to follow.

All previously open review threads have been addressed by the latest commits and are now resolved.

File: onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h [hunk 17]
```diff
@@ -73,20 +80,49 @@ Status CheckInputs(const T* input,
...
     transposed = true;
   }
-  int max_sequence_length = static_cast<int>(cos_cache_dims[0]);
-  int head_size = rotary_embedding_dim == 0 ? static_cast<int>(cos_cache_dims[1]) * 2
-                                            : static_cast<int>(hidden_size / num_heads);
+  int max_sequence_length = 0;
+  ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(cos_cache_dims[0], "max_sequence_length", max_sequence_length));
+  if (rotary_embedding_dim == 0) {
+    int cache_width = 0;
+    ORT_RETURN_IF_ERROR(detail::NarrowNonNegativeToInt32(cos_cache_dims[1], "cache_width", cache_width));
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/rotary_embedding_helper.h:117]: When `rotary_embedding_dim == 0`, this overwrites `head_size` with `cos_cache_dims[1] * 2`. For rank-4 input we already had the real input head size from `input_dims[3]`, and for rank-3 with a provided `num_heads` we need `num_heads * head_size` to match `hidden_size`. Deriving `head_size` from the cache makes the later check against `cos_cache_dims[1]` tautological, so a malformed cache can change `seq_stride`/`head_stride`/`batch_stride` and `rotary_embedding_dim`. For example, input `[1, 2, 1, 4]` with cache width `8` would make `RunRotaryEmbedding` process 16 elements per head against an output whose head dimension is only 4. Please preserve the input-derived head size where available, validate rank-3 consistency when `num_heads` is provided, and then check the cache width against that head size.

**Review by tianleiwu — CHANGES_REQUESTED** [review]
## Review Summary

The current head addresses the earlier overflow-helper and position arithmetic concerns well, but I found one remaining correctness issue in the contrib RotaryEmbedding validation. When `rotary_embedding_dim == 0`, the helper can derive `head_size` from the cache width instead of validating the cache against the input layout. That lets malformed cache shapes drive the runtime strides and row length used for output pointer arithmetic.

Requesting changes for the inline item below.

**Review by tianleiwu — DISMISSED** [review]

**Comment by tianleiwu** [comment]
LGTM. Please fix build pipelines

**Review by tianleiwu — APPROVED** [review]
